src/watch/fileWatchers.ts

```diff
@@ -72,15 +79,24 @@ export default class FileWatcher {
 		const handleWatchEvent = (event: string) => {
 			if (event === 'rename' || event === 'unlink') {
 				this.fsWatcher.close();
-				this.trigger();
+				this.trigger(id);
 				dispose();
 			} else {
-				// this is necessary because we get duplicate events...
-				const contents = fs.readFileSync(id, 'utf-8');
-				if (contents !== data) {
-					data = contents;
-					this.trigger();
+				let stats: fs.Stats;
+				try {
+					stats = fs.statSync(id);
+				} catch (err) {
+					if (err.code === 'ENOENT') {
+						if (mtime !== -1) {
+							mtime = -1;
+							this.trigger(id);
+						}
+						return;
+					}
+					throw err;
 				}
+				// debounce
+				if (+stats.mtime - mtime > 50) this.trigger(id);
```

Comment:
Could you explain in  a few words the idea here? So we ignore changes within 50ms of the previous change? Is this for the duplicate events mentioned in the replaced code or this is a certain scenario we try to solve here?

Reply:
In any case I must admit the new logic looks definitely more reliable 👍

Reply:
So when watchers emit file events at the same time (say save happens twice, or gets triggered as two events somehow), then you don't want to retrigger a whole rebuild. The debouncing approach just ignores events happening within 50ms of eachother, instead of having to read the whole source to determine. This then will also work with folder events too.

Reply:
Ok, but another situation might be that the file is saved first but then a moment later again because some automated process did some modification to it. In this case, the watcher might rerun with an outdated copy of the file (or not because there is still a setTimeout in the watcher, but there is no coded relation between the two).

Also you are doing this on a per-file basis. Wouldn't it make more sense to do the debouncing globally i.e. in `Watcher._makeDirty`? At the moment, the operation triggered by the watcher is a "set dirty" which is rather cheap to perform.

So how about this approach: We do not debounce here. Instead once `Watcher._makeDirty` is triggered, we start a timeout that is reset and restarted if `_makeDirty` is called again within this time. Only if the timeout completes without being interrupted do we trigger a new rollup run. That way, we can await arbitrarily long save operations including many files as long as there are no long pauses between the saves. I would assume that other tools handle this similarly and we no longer have the magic number.

Reply:
Also, now that we have TypeScript, it would be nice to replace the underscores in `Watcher` with proper `private` modifiers.